File: src/chapter01/q02/main.go
package main

import (
	"fmt"
	"os"
)

func reverseStr(s string) string {
	r := []rune(s)
	for i, j := 0, len(r)-2; i < j; i, j = i+1, j-1 {
		r[i], r[j] = r[j], r[i]
	}
	return string(r)
}

func main() {
	if reverseStr("abcd!") != "dcba!" {
		os.Exit(1)
	} else {
		fmt.Println("Ok")
	}
	if reverseStr("!") != "!" {
		os.Exit(1)
	} else {
		fmt.Println("Ok")
	}
}

File: src/chapter01/q02/main.py
import unittest

# NOTE use "!" to represent null


def reverse_str(s: str) -> str:
    # magic index !!!
    return s[-2::-1] + '!'


class TestCase(unittest.TestCase):
    def test_case1(self):
        self.assertEqual(reverse_str("abcd!"), "dcba!")

    def test_case2(self):
        self.assertEqual(reverse_str("!"), "!")


if __name__ == "__main__":
    unittest.main()
